### DISCUSSION TITLE
use18n only returns currentLanguageTag and labelBySupportedLanguageTag

### DISCUSSION THREAD

**Discussion by frodolfo** [keycloakify/keycloakify#204] [Q&A]
I tried to use use18n in the following:

```
const { kcContext } = getKcContext();

const i18n = useI18n({
    kcContext,
    extraMessages: JSON.stringify({
      en: {
        test: 'This is a test',
      },
      fr: {
        test: "C'est un test",
      },
      es: {
        test: 'Esto es una prueba',
      },
    }),
  });
  
  const kcProps = {
    i18n,
    ...defaultKcProps,
  };
```

And then I pass i18n as a prop to a component.  

`
<MyComponent {...{ kcContext, ...kcProps }} />
`

Within that **MyComponent** I tried to extract **msg** and **msgStr** but they're null

```
const { kcContext, i18n } = props;
const { msg, msgStr } = i18n;
```

i18n only has the following

```
{
   "currentLanguageTag":"en",
   "labelBySupportedLanguageTag":{}
}
```

It seems like **useI18n** isn't giving me anything at all.  What am I missing?

**Comment by garronej** [comment]
`i18n` resources are fetched dynamically so we don't bundle every translation for every languages.  

You have to:  
https://github.com/garronej/keycloakify-advanced-starter/blob/eee2f84f2e64e95d9958f475d026e3891d4be02b/src/KcApp/KcApp.tsx#L19-L22  

You should try TypeScript btw

**Reply by frodolfo** [reply]
@garronej  - Thanks for responding.  
Our code base is still largely in javascript so I'm kinda stuck with it until our team decide it's time to convert everything to typescript .  Anyway, I saw the return null line in the sample app when i18n hasn't been set yet, but I cannot do that because the page I'm rendering ends up being blank.  
I guess the real question I should ask is when does the locale actually get set so that I can start utilizing `msg` and `msgStr`?  
I understand that in the Realm setting on Keycloak, I have to enable localization and such, but when does Keycloak tell the frontend/keycloakify that there is a locale.  Second question is how do I access the default keys in `messages_xx.properties`?

**Reply by garronej** [reply]
> Anyway, I saw the return null line in the sample app when i18n hasn't been set yet, but I cannot do that because the page I'm 
> rendering ends up being blank.  

To be sure we are on the same page, I'm returning `null` in the react component becuase the time required to fetch the local resource is usually so short that it's not worth bothering displaying a spinner.  
Now if the `i18n` stays `null` indefinately there is a problem.  Look in your console, you should have a 404.  

> I guess the real question I should ask is when does the locale actually get set so that I can start utilizing msg and msgStr?  

It get sets automatically as soon as possible.  

> I understand that in the Realm setting on Keycloak, I have to enable localization and such  

I mean, it's not even requiered to do so. If you don't it will fallback to english.  

>  but when does Keycloak tell the frontend/keycloakify that there is a locale.  

Here: https://github.com/InseeFrLab/keycloakify/blob/fa72a29999d09c8a2005efda0e103c9691f22f56/src/lib/i18n/index.tsx#L89  

>  Second question is how do I access the default keys in messages_xx.properties?  

You don't, [the default translation are distributed by Keycloakify](https://github.com/InseeFrLab/keycloakify/blob/fa72a29999d09c8a2005efda0e103c9691f22f56/src/lib/i18n/index.tsx#L94-L142) and if you want to overwrite the default you do it like so: https://github.com/garronej/keycloakify-advanced-starter/blob/eee2f84f2e64e95d9958f475d026e3891d4be02b/src/KcApp/i18n.ts

**Reply by frodolfo** [reply]
@garronej  Here's the basic hierarchy of my component:

```
index.js 
   |
   `---- kcApp.js
           |
           `--- MyComponent.js
```
1) `index.js` - grabs `kcContext` from getKcContext() then passes it to `kcApp.js` as a prop along with `defaultKcProps`
2) `kcApp.js` - this is where `usei18n()` is invoked and sets the value of `i18n` along with `extraMessages` just like in your reference above. `i18n` is grouped with `kcProps` along with `defaultKcProps`.  `kcProps` and `kcContext` are passed to `MyComponent.js` as props
3) `MyComponent.js` - this is where I try to grab `msg` and `msgStr` from `i18n`, but i18n only has the following:
```
{
    "currentLanguageTag":"en-US",
    "labelBySupportedLanguageTag": {
         "en-CA":"en-CA",
         "en-US":"en-US",
         "es-PR":"es-PR",
         "en":"English",
         "es-US":"es-US",
         "es-CL":"es-CL",
         "fr":"Français",
         "es":"Español",
         "fr-CA":"fr-CA"
      }
}
```
`labelBySupportedLanguageTag` is no longer an empty object because I configured Keycloak to have some localization values.  There is no 404 error in the network tab of the dev tool.  No errors in the console either.

So when I load the MyComponent page, I don't get any of the other properties that i18n is supposed to have.  I don't know if there's a setup or setting that I'm missing.

The other challenge I have is we have to account for a more region-specifc locale (i.e. Spanish (Chile) and French (Canada)), so we cannot just go with `es` or `fr` as the translations vary.

**Reply by garronej** [reply]
>  this is where I try to grab msg and msgStr from i18n, but i18n only has the following  

This is impossible.  
https://github.com/InseeFrLab/keycloakify/blob/fa72a29999d09c8a2005efda0e103c9691f22f56/src/lib/i18n/index.tsx#L146-L174  

> The other challenge I have is we have to account for a more region-specifc locale (i.e. Spanish (Chile) and French (Canada)), so 
> we cannot just go with es or fr as the translations vary.

This is shouldn't be a problem.  
You can defnine arbitrary local in the Keycloak admin panel:  
<img width="1016" alt="image" src="https://user-images.githubusercontent.com/6702424/201450089-733b2e81-03b6-4f53-8605-501d7a51790b.png">

And provide translation for this arbitrary language.  

<img width="634" alt="image" src="https://user-images.githubusercontent.com/6702424/201450141-8d5e9c00-4897-4125-b80c-db378e203cec.png">

What isn't supported though is to tell to what language fallback to when a specific translation is missing. It always fallback to english.  When it would be best that Chile would fall back to Spanish....

**Reply by frodolfo** [reply]
> > this is where I try to grab msg and msgStr from i18n, but i18n only has the following
> 
> This is impossible.
> 
> https://github.com/InseeFrLab/keycloakify/blob/fa72a29999d09c8a2005efda0e103c9691f22f56/src/lib/i18n/index.tsx#L146-L174
> 

I'm starting to think that there might be a problem with the way we build the theme.  Maybe you can tell me if there's a problem in this sequence:
1) Invoke `yarn keycloak`
2) Create a `dist` folder (referenced in docker-compose)
3) Copy everything in `./build_keycloak/src/main/resources/theme/rider-portal/login/` into `dist`
4) Make a `messages` directory within `dist/login`
5) Copy `messages_en.properties` into this location (I know we're not supposed to modify the properties file, which is why I'm trying to get i18n working) -- this step will be omitted later on
6) Run Docker

Does step 3 pose a problem?


> > The other challenge I have is we have to account for a more region-specifc locale (i.e. Spanish (Chile) and French (Canada)), so
> > we cannot just go with es or fr as the translations vary.
> 
> This is shouldn't be a problem. You can defnine arbitrary local in the Keycloak admin panel: <img alt="image" width="1016" src="https://user-images.githubusercontent.com/6702424/201450089-733b2e81-03b6-4f53-8605-501d7a51790b.png">
> 
> And provide translation for this arbitrary language.
> 
> <img alt="image" width="634" src="https://user-images.githubusercontent.com/6702424/201450141-8d5e9c00-4897-4125-b80c-db378e203cec.png">
> 

This is exactly what I did, but the `extraLanguages` isn't coming up at all.

> What isn't supported though is to tell to what language fallback to when a specific translation is missing. It always fallback to english. When it would be best that Chile would fall back to Spanish....

I'll make a note of this.  Thanks again.

**Reply by garronej** [reply]
> I'm starting to think that there might be a problem with the way we build the theme. Maybe you can tell me if there's a problem in this sequence:  

😰  

When you run `yarn keycloak` there is a `.jar` file generated in `build_keycloak/target` that's it. That's whay you need to import in Keycloak.  

If you want to test locally, just start Docker and run the command that is displayed to you after building your theme:  
<img width="784" alt="image" src="https://user-images.githubusercontent.com/6702424/201450999-05c90ea8-a938-4ad4-abd0-6c071dc84af4.png">

> This is exactly what I did, but the extraLanguages isn't coming up at all.  

It's normal, unless it's the language of your browser you have to call `i18n.changeLocal(`foo-bar-baz`)`, you can also test using mocks.  

https://github.com/garronej/keycloakify-advanced-starter/blob/eee2f84f2e64e95d9958f475d026e3891d4be02b/src/KcApp/kcContext.ts#L19-L22

**Reply by frodolfo** [reply]
> <img alt="image" width="634" src="https://user-images.githubusercontent.com/6702424/201450141-8d5e9c00-4897-4125-b80c-db378e203cec.png">

BTW, Is it important to have double-quotes around `extraLanguages` like a proper JSON structure?  My VS Code linter keeps stripping them off.

**Reply by garronej** [reply]
> BTW, Is it important to have double-quotes around extraLanguages like a proper JSON structure? My VS Code linter keeps stripping them off.  

No, it's juste my coding style.  
Declaring object JSON style makes it easyer to tell what is a type declaration and whant is an Object instantiation.

**Reply by frodolfo** [reply]
> > I'm starting to think that there might be a problem with the way we build the theme. Maybe you can tell me if there's a problem in this sequence:
> 
> 😰
> 
> When you run `yarn keycloak` there is a `.jar` file generated in `build_keycloak/target` that's it. That's whay you need to import in Keycloak.
> 

We're in a unique predicament because another team is in charge of setting up the Keycloak instance.  My team supplies the theme.  We have different teams with their own theme and only one Keycloak instance, so drop the theme the team responsible for our Keycloak instance picks up our theme as part of their build process.  Any other way around this or would we need to work with them and have them consume our generated JAR file instead.

> If you want to test locally, just start Docker and run the command that is displayed to you after building your theme: <img alt="image" width="784" src="https://user-images.githubusercontent.com/6702424/201450999-05c90ea8-a938-4ad4-abd0-6c071dc84af4.png">

We use the other team's Makefile to build and run Keycloak in a Docker container locally.  We're currently using Keycloak 19

**Reply by frodolfo** [reply]
So it looks like there are several themes in `/opt/keycloak/themes/`.  If we just copy the generatd JAR file into `/opt/keycloak/providers`, would these other themes be disregarded by Keycloak?   This is one concern I have because on our Administrator Console, we select these themes from the Realm's and clients' settings.  I suppose I can try it out and find out myself.

**Reply by garronej** [reply]
No, they won't be disregarded dont worry.  
You will be able to select the theme you want on a realm basis or on a client basis.

**Reply by frodolfo** [reply]
@garronej - okay I was able to just copy the JAR file in the `build_keycloak/target` directory to the `providers` directory of our Keycloak instance and the other themes remain accessible.  I verified that our theme is in the dropdown, and I was able to bring up our theme.  However, the `i18n` constant I have defined still only has the following:

```
"i18n":{
      "currentLanguageTag":"en",
      "labelBySupportedLanguageTag":{
         "en-CA":"en-CA",
         "es-PR":"es-PR",
         "en":"English",
         "es-US":"es-US",
         "es-CL":"es-CL",
         "fr":"Français",
         "es":"Español"
      }
   },
```

What else could possibly be missing?

**Reply by garronej** [reply]
> However, the i18n constant I have defined still only has the following  

I'm positive, this is imposible. Look at [the code](https://github.com/InseeFrLab/keycloakify/blob/main/src/lib/i18n/index.tsx#L145-L176).  

What's probably happening is that you are logging things like `console.log(JSON.stringify(i18n, null, 2))` and functions got stipped away.  

<img width="422" alt="image" src="https://user-images.githubusercontent.com/6702424/201472925-f8931354-2a48-4df9-8755-2e5b753f2537.png">

<img width="179" alt="image" src="https://user-images.githubusercontent.com/6702424/201472936-b58a0ea1-47af-4975-9d20-f41f0331addb.png">

**Reply by frodolfo** [reply]
I'll simplify how things were set up here.  I only have this console command:

```
const i18n = useI18n({
    kcContext,
    extraMessages: JSON.stringify({
      en: {
        test: 'This is a test',
      },
      fr: {
        test: "C'est un test",
      },
      es: {
        test: 'Esto es una prueba',
      },
    }),
  });
  
const kcProps = {
    i18n,
    ...props,
};

======

<MyComponent {...{ kcContext, ...kcProps }} />;

======
MyComponent = (props) => { 
const { kcContext, i18n } = props;

console.log(`i18n: ${JSON.stringify(i18n)}`);
}
```

If I also console.log `kcContext` in the same manner, I get a ton of stuff back.  I know the test passes so I'm pulling my hair as to why this is the case with the value of i18n.

BTW, I truly appreciate your patience and regular feedback.

**Reply by garronej** [reply]
> If I also console.log kcContext in the same manner, I get a ton of stuff back. I know the test passes so I'm pulling my hair as to why 
> this is the case with the value of i18n.  

Ha ha! I told you in my previous reply why it was the case, you just confirmed I was right. `JSON.stringify` stripes away functions, functions aren't serializable.  

Try this, you'll see  

```diff
- console.log(`i18n: ${JSON.stringify(i18n)}`);
+ console.log(`i18n: `, i18n);
```  

> BTW, I truly appreciate your patience and regular feedback.  

NP you are welcome.

**Reply by frodolfo** [reply]
> ```diff
> - console.log(`i18n: ${JSON.stringify(i18n)}`);
> + console.log(`i18n: `, i18n);
> ```

I've gotten used to always interpolating the variables I want to inspect that I forgot about using console.log this way.  The interpolation just gives me `[Object object]` all the time, and I always resort to just using `JSON.stringify()` when I want to see what's in the object.  Thanks a lot!

**Reply by frodolfo** [reply]
Got everything working.  Thanks again, Joseph.

**Reply by garronej** [reply]
Great!

**Reply by frodolfo** [reply]
@garronej - one other question: can I have the `extraMessages` structured like this:

```
"en": {
  "pages": {
    "login": {
      "titleSignIn": "Sign In",
      "usernameOrEmail": "Email or Username",
      "labelSignin": "Sign In",
      "labelCreateAccount": "Create Account",
      "messageLoginTimeout": "Your login attempt timed out. Login will start from the beginning.  Please try again",
      "messageSuccessFulLogout": "You've successfully signed out.",
      "messageForgotPassword": "If you have an account, a recovery link has been sent to your email.  Please check your spam junk folder if you do not receive it, or contact your transit agency for further assistance."
    }
  }
 }
```

If so, how would I access the login-specific keys using `msgStr()`?

**Reply by garronej** [reply]
No, the keys aren't scoped.

**Reply by garronej** [reply]
Hello @frodolfo,  
Sorry, I think the error is out of the scope of what you are seeing me.  
Try to locate the exact line that throws by surrounding the suspects with `console.log("before"); console.log("after");` if you only see the `before`, you got your line.

**Reply by frodolfo** [reply]
Oops, I had deleted my previous message because I figured it out, but I didn't realize you had already replied to it.  One of the child components was missing i18n in the props.  Thanks again.

**Reply by garronej** [reply]
NP you are welcome
